core/src/main/java/com/linecorp/armeria/server/http/auth/HttpAuthService.java

```diff
@@ -83,18 +91,30 @@ protected HttpResponse onSuccess(ServiceRequestContext ctx, HttpRequest req) thr
      * Invoked when {@code req} is failed. By default, this method responds with the
      * {@link HttpStatus#UNAUTHORIZED} status.
      */
-    protected HttpResponse onFailure(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+    protected HttpResponse onFailure(ServiceRequestContext ctx, HttpRequest req, @Nullable Throwable t)
+            throws Exception {
+        logger.warn("Unexpected exception during authorization.", t);
         final DefaultHttpResponse res = new DefaultHttpResponse();
         res.respond(HttpStatus.UNAUTHORIZED);
         return res;
     }
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        if (authorize(req.headers())) {
-            return onSuccess(ctx, req);
-        } else {
-            return onFailure(ctx, req);
-        }
+        DeferredHttpResponse res = new DeferredHttpResponse();
+
+        authorize(req, ctx).whenCompleteAsync((result, t) -> {
```

Comment:
Hmm, could you explain why you switched from `whenComplete` to `whenCompleteAsync`?